File: backend/app/schemas/passport.py
"""Pydantic schemas for Verified Skill Passport and Skill Evidence domain."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.skill_evidence import EvidenceStatus, EvidenceType
from app.models.verified_skill import VerificationMethod, VerificationStatus


class SkillEvidenceCreate(BaseModel):
    """Schema for candidate manually creating skill evidence (e.g. certification)."""

    skill_id: uuid.UUID
    evidence_type: EvidenceType = Field(
        default=EvidenceType.CERTIFICATION,
        description="Type of evidence (e.g., CERTIFICATION, CANDIDATE_DECLARATION)",
    )
    title: str = Field(
        ..., min_length=1, max_length=255, description="Evidence title or credential name"
    )
    description: str | None = Field(default=None, description="Detailed description or notes")
    evidence_url: str | None = Field(
        default=None, max_length=500, description="Link or verification URL"
    )
    issued_at: datetime | None = Field(default=None, description="Issuing date")
    completed_at: datetime | None = Field(default=None, description="Completion date")
    meta: dict[str, Any] | None = Field(
        default=None, description="Structured attributes (e.g., issuer, credential ID)"
    )


class SkillEvidenceRead(BaseModel):
    """Schema for reading a skill evidence item."""

    id: uuid.UUID
    candidate_id: uuid.UUID
    skill_id: uuid.UUID
    skill_name: str | None = None
    evidence_type: EvidenceType
    source_id: uuid.UUID | None = None
    title: str
    description: str | None = None
    evidence_url: str | None = None
    issued_at: datetime | None = None
    completed_at: datetime | None = None
    meta: dict[str, Any] | None = None
    status: EvidenceStatus
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class VerifiedSkillItem(BaseModel):
    """Canonical representation of a skill within the candidate passport."""

    skill_id: uuid.UUID
    skill_name: str
    skill_slug: str
    category: str
    skill_type: str
    status: VerificationStatus
    verification_method: VerificationMethod
    verification_score: float | None = None
    verified_at: datetime | None = None
    expires_at: datetime | None = None
    verification_summary: str
    evidence_count: int = 0
    strongest_evidence_type: EvidenceType | None = None
    evidence_items: list[SkillEvidenceRead] = []

    model_config = ConfigDict(from_attributes=True)


class PassportCandidateSummary(BaseModel):
    """High-level summary of candidate profile for passport display."""

    candidate_id: uuid.UUID
    user_id: uuid.UUID
    full_name: str
    headline: str | None = None
    current_role: str | None = None
    location_city: str | None = None
    location_state: str | None = None

    model_config = ConfigDict(from_attributes=True)


class PassportStats(BaseModel):
    """Statistical breakdown of verification metrics."""

    total_skills: int = 0
    verified_skills: int = 0
    unverified_skills: int = 0
    expired_skills: int = 0
    total_evidence_items: int = 0
    verification_coverage_pct: float = 0.0


class CandidatePassportResponse(BaseModel):
    """Full candidate-owned Verified Skill Passport response."""

    candidate: PassportCandidateSummary
    stats: PassportStats
    skills: list[VerifiedSkillItem]
    last_recalculated_at: datetime | None = None
    share_token: str | None = None
    is_share_enabled: bool = False


class PublicPassportResponse(BaseModel):
    """Publicly shareable Verified Skill Passport profile."""

    candidate: PassportCandidateSummary
    stats: PassportStats
    skills: list[VerifiedSkillItem]
    shared_at: datetime | None = None


class PassportShareToggleRequest(BaseModel):
    """Payload to enable or disable public passport sharing."""

    is_enabled: bool = Field(..., description="Whether public sharing is enabled")


class PassportShareResponse(BaseModel):
    """Public sharing token configuration response."""

    share_token: str
    is_enabled: bool
    share_url: str | None = None

File: frontend/src/types/passport.ts
/**
 * Verified Skill Passport and Evidence TypeScript interfaces.
 */

export type EvidenceType =
  | "COURSE_COMPLETION"
  | "CANDIDATE_DECLARATION"
  | "RESUME_EXTRACTION"
  | "CERTIFICATION"
  | "ASSESSMENT";

export type EvidenceStatus = "VALID" | "REVOKED" | "EXPIRED";

export type VerificationStatus = "VERIFIED" | "UNVERIFIED" | "EXPIRED" | "REJECTED";

export type VerificationMethod =
  | "COURSE_COMPLETION"
  | "CERTIFICATION"
  | "ASSESSMENT"
  | "CANDIDATE_DECLARATION"
  | "RESUME_EXTRACTION"
  | "NONE";

export interface SkillEvidenceRead {
  id: string;
  candidate_id: string;
  skill_id: string;
  skill_name?: string | null;
  evidence_type: EvidenceType;
  source_id?: string | null;
  title: string;
  description?: string | null;
  evidence_url?: string | null;
  issued_at?: string | null;
  completed_at?: string | null;
  meta?: Record<string, unknown> | null;
  status: EvidenceStatus;
  created_at: string;
  updated_at: string;
}

export interface SkillEvidenceCreate {
  skill_id: string;
  evidence_type: EvidenceType;
  title: string;
  description?: string;
  evidence_url?: string;
  issued_at?: string;
  completed_at?: string;
  meta?: Record<string, unknown>;
}

export interface VerifiedSkillItem {
  skill_id: string;
  skill_name: string;
  skill_slug: string;
  category: string;
  skill_type: string;
  status: VerificationStatus;
  verification_method: VerificationMethod;
  verification_score?: number | null;
  verified_at?: string | null;
  expires_at?: string | null;
  verification_summary: string;
  evidence_count: number;
  strongest_evidence_type?: EvidenceType | null;
  evidence_items: SkillEvidenceRead[];
}

export interface PassportCandidateSummary {
  candidate_id: string;
  user_id: string;
  full_name: string;
  headline?: string | null;
  current_role?: string | null;
  location_city?: string | null;
  location_state?: string | null;
}

export interface PassportStats {
  total_skills: number;
  verified_skills: number;
  unverified_skills: number;
  expired_skills: number;
  total_evidence_items: number;
  verification_coverage_pct: number;
}

export interface CandidatePassportResponse {
  candidate: PassportCandidateSummary;
  stats: PassportStats;
  skills: VerifiedSkillItem[];
  last_recalculated_at?: string | null;
  share_token?: string | null;
  is_share_enabled: boolean;
}

export interface PublicPassportResponse {
  candidate: PassportCandidateSummary;
  stats: PassportStats;
  skills: VerifiedSkillItem[];
  shared_at?: string | null;
}

export interface PassportShareToggleRequest {
  is_enabled: boolean;
}

export interface PassportShareResponse {
  share_token: string;
  is_enabled: boolean;
  share_url?: string | null;
}
